Sources/SecureXPC/Server/XPCServer.swift

```diff
@@ -236,19 +283,64 @@ public class XPCServer {
         }
     }
     
-    private func handleMessage(connection: xpc_connection_t, message: xpc_object_t, reply: inout xpc_object_t?) throws {
-        let request = try Request(dictionary: message)
+    private func handleMessage(connection: xpc_connection_t, message: xpc_object_t) {
+        let request: Request
+        do {
+            request = try Request(dictionary: message)
+        } catch {
+            var reply = xpc_dictionary_create_reply(message)
+            self.handleError(error, expectingOtherError: false, connection: connection, reply: &reply)
+            return
+        }
+        
         guard let handler = self.routes[request.route] else {
-            throw XPCError.routeNotRegistered(request.route.pathComponents)
+            let error = XPCError.routeNotRegistered(request.route.pathComponents)
+            var reply = xpc_dictionary_create_reply(message)
+            self.handleError(error, expectingOtherError: false, connection: connection, reply: &reply)
+            return
         }
-        try handler.handle(request: request, reply: &reply)
         
-        // If a dictionary reply exists, then the message expects a reply to be sent back
-        if let reply = reply {
-            xpc_connection_send_message(connection, reply)
+        if let handler = handler as? XPCHandlerSync {
+            var reply = xpc_dictionary_create_reply(message)
+            do {
+                try handler.handle(request: request, reply: &reply)
+                
+                // If a dictionary reply exists, then the message expects a reply to be sent back
+                if let reply = reply {
+                    xpc_connection_send_message(connection, reply)
+                }
+            } catch {
+                self.handleError(error, expectingOtherError: true, connection: connection, reply: &reply)
+            }
+            
+        } else if #available(macOS 10.15.0, *), let handler = handler as? XPCHandlerAsync {
+            Task {
+                var reply = xpc_dictionary_create_reply(message)
+                do {
+                    try await handler.handle(request: request, reply: &reply)
+                    
+                    // If a dictionary reply exists, then the message expects a reply to be sent back
+                    if let reply = reply {
+                        xpc_connection_send_message(connection, reply)
+                    }
+                } catch {
+                    self.handleError(error, expectingOtherError: true, connection: connection, reply: &reply)
+                }
+            }
+        } else {
+            fatalError("Server handler for route \(request.route.pathComponents) cannot be run: \(handler)")
```

Comment:
Could you clarify this case?

By my reading, it happens if you have a `XPCHandlerAsync` registered, but don't pass the `#available(macOS 10.15.0, *)` check?

Reply:
Yeah. It shouldn't be possible, but seemed prudent to explicitly fail if it was encountered. I can improve the error message?

Reply:
Just explain why it doesn't makes sense, along the lines of "The route wasn't an sync route, even that's the only kind that should be possible for this system"